base/channels.jl

```diff
@@ -457,3 +457,24 @@ function iterate(c::Channel, state=nothing)
 end
 
 IteratorSize(::Type{<:Channel}) = SizeUnknown()
+
+"""
+    threaded_foreach(f, channel::Channel; ntasks=Threads.nthreads(), async=true)
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` Tasks spawned by `Threads.@spawn`.
+
+If `async` is `false`, this function will `wait` for all internally spawned tasks
+to complete before returning.
+"""
+function threaded_foreach(f, channel::Channel; ntasks=Threads.nthreads(), async=true)
```

Comment:
Does it make sense to use the name `Threads.foreach`, like we have `Iterators.filter`?

Reply:
ref: @vchuravy suggested to use C++ -like policy-based interface https://github.com/JuliaLang/julia/pull/34185#issuecomment-568878123

Reply:
> Does it make sense to use the name Threads.foreach, like we have Iterators.filter?

That would be nice. 

One tricky detail (but not an actual problem) is that writing code that uses `Threads.@spawn` from within `Threads` is pretty wonky right now because the macro depends on functions that aren't defined until later (in `tasks.jl`) and `Channel`s also aren't defined until later. Clearly can churn things around to be defined in the right order but might involve some bikeshedding to untangle the dependencies

Reply:
Ah, I didn't know that.  I guess one way to work around the problem is to just declare the "stub" function like this?

```julia
module Threads
function foreach end
...
end

# later
function Threads.foreach(f, channel::Channel)
    ...
end
```

Reply:
I pushed the above change, since I like `Threads.foreach` and the stub approach is way less hassle/churn than actually shuffling code :p 

Folks should, of course, feel free to continue bikeshedding the name.